#112: Add server-client protocol and feature matrix (#130)
eclipse-glsp/glsp: PROTOCOL.md

<details open><summary>Code</summary>

```typescript
class SetMarkersAction implements Action {
    /**
     * The kind of the action.
     */
    public readonly kind = "setMarkers";

    /**
     * The list of markers that has been requested by the `RequestMarkersAction`.
     */
    public readonly markers: Marker[];
}
```
</details>

### 2.11.3. DeleteMarkersAction

To remove markers for elements a client or server may send a `DeleteMarkersAction` with all markers that should be removed.

<details open><summary>Code</summary>

```typescript
class DeleteMarkersAction implements Action {
    /**
     * The kind of the action.
     */
    public readonly kind = "deleteMarkers";

    /**
     * The list of markers that should be deleted.
     */
    public readonly markers: Marker[];
}
```
</details>

## 2.12. Element Navigation

GLSP makes no assumption about the type of navigation a user may want to perform. Thus a generic infrastructure is provided that the client and server can use to implement specific navigation types, e.g., navigation to documentation, implementation, etc. The type of navigation is identified by the `targetTypeId`.

> A client may request the targets for a specific type of navigation by querying the server to which the server will respond with a set of navigation targets. A `NavigationTarget` identifies the object we want to navigate to via it's uri and may further provide a label to display for the client. Additionally, generic  arguments may be used to to encode any domain- or navigation type-specific information.

Reviewer: ```suggestion
A client may request the targets for a specific type of navigation by querying the server to which the server will respond with a set of navigation targets. A `NavigationTarget` identifies the object we want to navigate to via its uri and may further provide a label to display for the client. Additionally, generic arguments may be used to to encode any domain- or navigation type-specific information.
```

```typescript
class RequestEditValidationAction implements RequestAction<SetEditValidationResultAction> {
    /**
     * The kind of the action.
     */
    public readonly kind = "requestEditValidation";

    /**
     * Context in which the text is validated, e.g., 'label-edit'.
     */
    public readonly contextId: string;

    /**
     * Model element that is being edited.
     */
    public readonly modelElementId: string;

    /**
     * Text that should be considered for the model element.
     */
    public readonly text: string;
}
```
</details>

### 2.17.2. SetEditValidationResultAction

Response to a `RequestEditValidationAction` containing the validation result for applying a text on a certain model element.

<details open><summary>Code</summary>

```typescript
class SetEditValidationResultAction implements ResponseAction {
    /**
     * The kind of the action.
     */
    public readonly kind = "setEditValidationResult";

    /**
     * Validation status.
     */
    public readonly status: ValidationStatus;

    /*
     * Additional arguments for custom behavior.
     */
    public readonly args?: Args;
}
```
</details>

### 2.17.3. ApplyLabelEditOperation

> A very common use case in domain models is the support of labels that display textual information to the user. For instance, the `SGraph` model of Sprotty has support for labels that can be attached to a node, edge, or port, and that contain some text that is rendered in the view. To apply new text to such a label element the client may send a `ApplyLabelEditOperation` to the server.

Reviewer: ```suggestion
A very common use case in domain models is the support of labels that display textual information to the user. For instance, the `SGraph` model of Sprotty has support for labels that can be attached to a node, edge, or port, and that contain some text that is rendered in the view. To apply new text to such a label element the client may send an `ApplyLabelEditOperation` to the server.
```

<details open><summary>Code</summary>

```typescript
class ApplyLabelEditOperation implements Operation {
    /**
     * The kind of the action.
     */
    public readonly kind = "applyLabelEdit";

    /**
     * Identifier of the label model element.
     */
    public readonly labelId: string;

    /**
     * Text that should be applied on the label.
     */
    public readonly text: string;
}
```
</details>

## 2.18. Clipboard

> In GLSP the clipboard needs to be managed by the client but the conversion from data to a clipboard-compatible format is handled by the server. By default, we use `application/json` as exchange format.

Reviewer: ```suggestion
In GLSP the clipboard needs to be managed by the client but the conversion from the selection to be copied into a clipboard-compatible format is handled by the server. By default, GLSP use `application/json` as exchange format.
```

    /**
     * The current editor context.
     */
    public readonly editorContext: EditorContext;
}
```
</details>

### 2.20.2. SetContextActions

The `SetContextActions` is the response to a `RequestContextActions` containing all actions for the queried context.

<details open><summary>Code</summary>

```typescript
class SetContextActions implements ResponseAction {
    /**
     * The kind of the action.
     */
    public readonly kind = "setContextActions";

    /**
     * The actions available in the queried context.
     */
    public readonly actions: LabeledAction[];

    /**
     * Custom arguments.
     */
    public readonly args: Args?;
}
```
</details>

### 2.20.3. Context Menu

The context menu is an overlay that is triggered by a right click from the user. The menu may be filled with actions from the client but may also be filled with actions from the server. If server actions are to be used, the client needs to send a `RequestContextActions` action with context id `context-menu` and handle the returned actions from the `SetContextActions` response accordingly, e.g., rendering them in a context menu.

### 2.20.4. Command Palette

> The command palette is an "auto-complete" widget that is triggered the user hitting `ctrl+space`. The menu may be filled with actions from the client but may also be filled with actions from the server. If server actions are to be used, the client needs to send a `RequestContextActions` action with context id `command-palette` and handle the returned actions from the `SetContextActions` response accordingly, e.g., rendering them in a auto-complete widget.

Reviewer: ```suggestion
The command palette is an "auto-complete" widget that is triggered when the user hits `Ctrl+Space`. The menu may be filled with actions from the client but may also be filled with actions from the server. If server actions are to be used, the client needs to send a `RequestContextActions` action with context id `command-palette` and handle the returned actions from the `SetContextActions` response accordingly, i.e., rendering them in a auto-complete widget.
```
